### ISSUE TITLE
[API Proposal]: Change `System.Numerics.IPowerFunctions`

### ISSUE THREAD

**Issue by mnelsonwhite** [dotnet/runtime#85707]
### Background and motivation

Currently, if I want to support a `Pow` operation for some generic type, I am not really able to give the exponent a sensible value. It is great the current numerics API supports both an additive and multiplicative identity, but I am not sure it is correct to get a square of a generic constrained to  `T: IPowerFunctions<T,T>, IAdditionOperators<T,T,T>, IMultiplicativeIdentity<T>` by doing `T.Pow(value, T.MultiplicativeIdentity + T.MultiplicativeIdentity)`

For this reason I think an exponent type of `double` is workable for all cases and makes mathematical sense.
Alternatively allow for `IPowerFunctions<TSelf, TExponent>` so that `TExponent` can be constrained to `double` or other specific numeric types.

If I have a field of `F` it doesn't make much sense to raise a value of `F` to the power of `F` as `F^F` since the exponent should be a real number describe the number of times `F` is multiplied by itself, e.g. `F.F.F.F = F^4`  and `4` may not be represented at all within the set of the field `F` since it is within the set of `R` and `Z`.

### API Proposal

```csharp
namespace System.Numerics
{
    //
    // Summary:
    //     Defines support for power functions.
    //
    // Type parameters:
    //   TSelf:
    //     The type that implements this interface.
    public interface IPowerFunctions<TSelf>
        : INumberBase<TSelf>
        , IEquatable<TSelf>
        , IFormattable
        , IParsable<TSelf>
        , ISpanFormattable
        , ISpanParsable<TSelf>
        , IAdditionOperators<TSelf, TSelf, TSelf>
        , IAdditiveIdentity<TSelf, TSelf>
        , IDecrementOperators<TSelf>
        , IDivisionOperators<TSelf, TSelf, TSelf>
        , IEqualityOperators<TSelf, TSelf, bool>
        , IIncrementOperators<TSelf>
        , IMultiplicativeIdentity<TSelf, TSelf>
        , IMultiplyOperators<TSelf, TSelf, TSelf>
        , ISubtractionOperators<TSelf, TSelf, TSelf>
        , IUnaryNegationOperators<TSelf, TSelf>
        , IUnaryPlusOperators<TSelf, TSelf>
        where TSelf : IPowerFunctions<TSelf>?
    {
        //
        // Summary:
        //     Computes a value raised to a given power.
        //
        // Parameters:
        //   value:
        //     The value which is raised to the power of exponent.
        //
        //   exponent:
        //     The power to which value is raised.
        //
        // Returns:
        //     value raised to the power of exponent.
        static abstract TSelf Pow(TSelf value, double exponent);
    }
}
```

### API Usage

```csharp
public static class Example
{
    public T Square<T>(T value) where T : IPowerFunctions<T>
    {
        return T.Pow(2);
    }

    public T SquareRoot<T>(T value) where T : IPowerFunctions<T>
    {
        return T.Pow(0.5);
    }
}
```

### Alternative Designs

I suppose the below would also be suitable since it shouldn't break existing APIs and it should also allow the user to constrain `TExponent` 

```csharp
namespace System.Numerics
{
    //
    // Summary:
    //     Defines support for power functions.
    //
    // Type parameters:
    //   TSelf:
    //     The type that implements this interface.
    //   TExponent:
    //     The type that describes the exponent
    public interface IPowerFunctions<TSelf, TExponent>
        : INumberBase<TSelf>
        , IEquatable<TSelf>
        , IFormattable
        , IParsable<TSelf>
        , ISpanFormattable
        , ISpanParsable<TSelf>
        , IAdditionOperators<TSelf, TSelf, TSelf>
        , IAdditiveIdentity<TSelf, TSelf>
        , IDecrementOperators<TSelf>
        , IDivisionOperators<TSelf, TSelf, TSelf>
        , IEqualityOperators<TSelf, TSelf, bool>
        , IIncrementOperators<TSelf>
        , IMultiplicativeIdentity<TSelf, TSelf>
        , IMultiplyOperators<TSelf, TSelf, TSelf>
        , ISubtractionOperators<TSelf, TSelf, TSelf>
        , IUnaryNegationOperators<TSelf, TSelf>
        , IUnaryPlusOperators<TSelf, TSelf>
        where TSelf : IPowerFunctions<TSelf>?
        where TExponent : INumberBase<TExponent>
    {
        //
        // Summary:
        //     Computes a value raised to a given power.
        //
        // Parameters:
        //   value:
        //     The value which is raised to the power of value.
        //
        //   exponent:
        //     The power to which value is raised.
        //
        // Returns:
        //     value raised to the power of exponent.
        static abstract TSelf Pow(TSelf value, TExponent exponent);
    }

    //
    // Summary:
    //     Defines support for power functions.
    //
    // Type parameters:
    //   TSelf:
    //     The type that implements this interface.
    public interface IPowerFunctions<TSelf>
        : IPowerFunctions<TSelf, TSelf>
        where TSelf : IPowerFunctions<TSelf>?
        where TExponent : INumberBase<TExponent>
    {
    }
}
```
#### Alternative API Usage
```csharp
public static class Example
{
    public T Square<T>(T value) where T : IPowerFunctions<T, double>
    {
        return value.Pow(2);
    }

    public T SquareRoot<T>(T value) where T : IPowerFunctions<T, double>
    {
        return value.Pow(0.5);
    }
}
```

**Comment by tannergooding** [comment]
Thanks for the suggestion! The TL;DR; is that this was already considered as part of the overall and initial designs and was decided against.

> I think I understand why the API has been made this way, since we would expect a double back from a Pow operation on a double and likewise for an int

Yes, outside of conversion functions, we typically follow the principle of `T-in, T-out`. So if you take in an `int`, you should return an `int`. This avoids a number of issues that can crop up due to overload resolution and other breaking changes or ambiguities that can occur due to future versioning.

> I believe that the method signature should accept an exponent value of type double, accepting that not all types will be able to return a result for all exponent values

This would be limiting to the overall interface, such as if you were to define and implement a `Binary128` type that implements the IEEE 754 quadruple-precision binary floating-point specification and therefore has more range and precision than `double`.

Another example is `Complex.Pow(Complex, Complex)` which is not limited to the `power` being a `real number`.

It additionally would represent potential problems with less precise types (such as `float` or `Half`) where many values cannot be used as given. Similar would be the case with integral types where negative inputs cannot be used (and therefore likely must throw) or for fractional inputs which would result in truncation in most scenarios.

-------------------------------------------

> For this reason I think an exponent type of double is workable for all cases and makes mathematical sense.

The simplest way to respond to this is with the somewhat controversial statement that `computer math != real math`.

Consider for example that a `ring` includes `addition` and `multiplication` and must satisfy a few different axioms. One of these axioms is that of associativity: `(a + b) + c == a + (b + c)`. You'll find, however, that this axiom is violated in the context of certain `overflow` scenarios for `integers` and in general for `floating-point types` (such as `float`, `double`, `Half`, or even `decimal`) due to implicit rounding that must occur after each operation. Similar examples then exist when looking at the associativity or distributivity of `multiplication`. We also find general violations when looking at other relatively simple operations including computing the inverse of a value (`int.MinValue` does not have a representable inverse as `int`).

Because of these differences and because most types operate in a finite domain, we explicitly opted to not implement or expose the generic math interfaces using mathematical concepts like `groups`, `rings`, or `fields`.

Trying to define this functionality in such strict terms, even giving allowances for the subtle differences that exist, would have been overall too restrictive and would likely have caused confusion whether a given type is "correctly" meeting the implementation requirements for a given interface. It would have also significantly raised the barrier to entry for programmers without a background in math.

Instead, we opted to expose a set of core interfaces representing the programmatic APIs that are exposed and to provide only minimal guarantees around what they do. For example, we expose `IAdditionOperators` which simply defines `operator +` and `operator checked +`. While this is generally expected to compute the `sum` of two values, there is no strict requirement that it be used that way and developers "could" still define that to do other things (like concatenation of two strings).

We then have a set of slightly higher level interfaces that group these core interfaces together into slightly stronger contracts. For example, we expose `INumberBase` which actually places a constraint that binary `operator +` means addition, defines a properties for `One`, `Zero`, and even an `AdditiveIdentity`.

Such interface also exposes a number of APIs that express programmatic concepts such as "parsing" and "formatting" or account for other differences that are largely only relevant to programming. For example we have both `Max` and `MaxNumber` which accounts for the existence of `NaN` and likewise have `CreateChecked`, `CreateSaturating`, and `CreateTruncating` to account for the different kinds of overflow that may be desired. We also exposes some domain specific math concepts such as `IsPositive` and `IsNegative` with the explicit callout that `IsPositive() != !IsNegative()`, because there may be types (such as `Complex` numbers) that are neither positive nor negative.

In summary, these interfaces were all very carefully designed and considered both from the mathematical and programmatic perspective. They exist in a way that makes them powerful, flexible, and generally sensible to use with the types of shapes and patterns that programmers will encounter and deal with. This includes both from a correctness and from a performance perspective. While there are still some limitations, such as `IPowerFunctions` only taking in a single generic type parameter, the overall design allows for correct versioning and expansion while still achieving the core needs and scenarios that users need across all possible types without introducing additional errors or edge-cases that must be considered.

**Comment by mnelsonwhite** [comment]
@tannergooding what about the alternative design to allow APIs to accept a different and/or constrained type for the exponent? Would you accept at least that the exponent doesn't "need" the be the same type as the base type?

**Comment by tannergooding** [comment]
> Would that mean that generic squaring should be implemented with something like:

Squaring is a special case that is likely best implemented as `value * value`. `Pow` is a general-purpose function and will always have a bit more overhead than simply doing `value * value` unless all the right inlining and other optimizations kick in.

> Also, is there a recommended Create... method to choose for cases where you don't actually expect overflow to happen?

It depends on whether you want saturation or truncation. Truncation is typically going to be the cheaper of the two simply due to how most computers operate and it being the "default" for most hardware primitives.

> Would you accept at least that the exponent doesn't "need" the be the same type as the base type?

This ends up going against the principle of `T-in, T-out` and could itself cause ambiguities with the existing single generic version. -- The "core" interfaces representing the operators are special because they more represent the IL contract of the operators and so may be used to build other types of generic interface hierarchies beyond math/numerics.

The best thing to do is to convert to the preferred intermediate type, do the operation, and then convert to the preferred target type, particularly given that for non floating-point types most inputs will otherwise be invalid and result in exceptions being thrown due to overflow. They may also have much worse perf considerations. -- For `int32` you have (ignoring `x^1` and `1^x`) `48,036` valid input combinations for `Pow(int, int)`. However, of those `45,050` of them are only valid as `x^2`. That leaves `2,986` "other combinations of `Pow(x, y)` with `y` no greater than `30`. You see similar limitations for `ulong` and other integral types; where there isn't much benefit to exposing such `Pow` functionality on such types. `BigInteger` is the notable exception due to allowing up to 2.14 billion bits.
